File: Leetcode/2405 Optimal Partition of String.ts
const ALPHABET_SIZE = 26;

function partitionString(s: string): number {
    const lastIndex: number[] = new Array(ALPHABET_SIZE).fill(-1);
    let startIndex = -1, nPartitions = 0;
    for(let i = 0; i < s.length; i++) {
        const charIndex = getCharIndex(s[i]);
        if(lastIndex[charIndex] >= startIndex) {
            nPartitions++;
            startIndex = i;
        }
        lastIndex[charIndex] = i;
    }
    return nPartitions;
}

const SMALL_A_CHAR_CODE = 'a'.charCodeAt(0);

function getCharIndex(char: string): number {
    return char.charCodeAt(0) - SMALL_A_CHAR_CODE;
}

File: Leetcode/2405 Optimal Partition of String.py
class Solution:
    def partitionString(self, s):
        last_index = [-1] * 26
        start_index, n_partions = -1, 0
        for i, char in enumerate(s):
            char_index = get_char_index(char)
            if last_index[char_index] >= start_index:
                n_partions += 1
                start_index = i
            last_index[char_index] = i
        return n_partions


def get_char_index(char):
    return ord(char) - ord('a')
